digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala

```diff
@@ -777,10 +792,25 @@ class RoadAddressService(roadLinkService: RoadLinkService, eventbus: DigiroadEve
         val endPoints = GeometryUtils.geometryEndpoints(rl.geometry)
         pointCloud.exists(p => GeometryUtils.areAdjacent(p, endPoints._1) || GeometryUtils.areAdjacent(p, endPoints._2))
       }.map(rl => rl.linkId -> rl).toMap
-    val missingLinks = withDynSession {
-      RoadAddressDAO.getMissingRoadAddresses(connectedLinks.keySet)
+    val (missingLinks, roadAddresses) = if (newSession) {
+      withDynSession {
+        (RoadAddressDAO.getMissingRoadAddresses(connectedLinks.keySet),
+          RoadAddressDAO.fetchByLinkId(connectedLinks.keySet, includeFloating = true))
+      }
+    } else {
+      (RoadAddressDAO.getMissingRoadAddresses(connectedLinks.keySet),
+        RoadAddressDAO.fetchByLinkId(connectedLinks.keySet, includeFloating = true))
     }
-    missingLinks.map(ml => RoadAddressLinkBuilder.build(connectedLinks(ml.linkId), ml))
+    val builtMissing = missingLinks.map(ml => RoadAddressLinkBuilder.build(connectedLinks(ml.linkId), ml))
+    val remainingAddresses = roadAddresses.filterNot(ra => builtMissing.exists(_.linkId == ra.linkId))
+
+    val filteredAddresses = remainingAddresses.filter(ra => {
+      !GeometryUtils.withinTolerance(ra.geometry, connectedLinks(ra.linkId).geometry, MaxDistanceDiffAllowed)
+    }).map(fa => {
+      RoadAddressLinkBuilder.build(connectedLinks(fa.linkId), fa, floating = false, Some(connectedLinks(fa.linkId).geometry))
+    })
+
+    builtMissing ++ filteredAddresses.filter(_.floating)
```

Comment:
Doesn't .filter(_.floating) include all and thus is unnecessary since we are setting "floating = false" for all the links?

Reply:
Not necessarily, the floating flag inside the builder only affects the outcoming RoadLinkType (used by the UI to define the styles, -1 if I recall correctly).
Actual floating value is still defined by the road address in question.